enable autocast support for user-specified device_type and dtype

puririshi98 opened the pull request:
Fixes #{55374}
https://github.com/pytorch/pytorch/issues/55374

puririshi98 committed:
rebase commit

puririshi98 committed:
rebase commit

puririshi98 commented:
@ngimel  @mcarilli

puririshi98 committed:
rebase commit

puririshi98 committed:
Update init.cpp

puririshi98 committed:
Merge branch 'autocast_rebase' of https://github.com/puririshi98/pytorch into autocast_rebase

puririshi98 committed:
Update autocast_mode.py

puririshi98 committed:
Update autocast_mode.py

puririshi98 committed:
Update autocast_mode.py

puririshi98 commented:
This is what convergence behavior for various casting regimes looks like with Adam lr=1e-4 and a batchsize of 32 on a Linear model updated with 10,000 minibatches of data.

puririshi98 commented:
Things seem to be working fine when I build the branch and manually test it on a CUDA GPU dgx1v32g node. However, the CI is consistently failing on several tests in test_quantization and test_autograd because I made it such that if device_type=cuda is supplied by the user and cuda is not available on the machine, we throw an error, as per @ngimel suggestion. This causes a problem for the CI since these tests are run on a 2xlarge instance which does not seem to have a gpu therefor cuda.is_available() returns false and I throw the error that cuda is not available. These CI may have passed in the past since the original torch.cuda.amp.autocast simply threw a warning if it was not available, instead of erroring.

ngimel commented:
@puririshi98 I think we should preserve previous behavior of warning if cuda is not available (to avoid bc-breaking concerns), can you update the PR?

ngimel commented on torch/cuda/amp/autocast_mode.py:
```diff
@@ -1,146 +1,14 @@
-                return func(*args, **kwargs)
-        return decorate_autocast
 
+autocast = functools.partial(torch.autocast, device_type='cuda')
```
I think it makes sense to add `fast_dtype` kwarg here, defaulting it to float16, to enable dtype selection for `torch.cuda.amp.autocast` and preserve legacy behavior.

ngimel reviewed (commented):
Can you please add tests for torch.autocast? It looks like you are relying on `torch.cuda.amp.autocast` and `torch.cpu.amp.autocast` tests.

ezyang commented:
@ngimel I didn't look at this. shout if you need me to.

ngimel commented:
@ezyang this PR doesn't make #61804 worse, so after some UX and testing issues are resolved I intend to land it. #61804 is still a hi-pri issue though.

mcarilli commented on torch/autocast_mode.py:
```diff
@@ -0,0 +1,163 @@
+
+class autocast(object):
+    r"""
+    Instances of :class:`autocast` serve as context managers or decorators that
```
Please re-copy the [original docstring](https://github.com/pytorch/pytorch/pull/61002/files#diff-dac4bd53ced015c8810b3b02fc5c2ec6c2b0658c5090b4fbbd09c96bd45087d1L16) to make sure the formatting and newlines match exactly, as the newlines _will_ affect the rendered paragraphs.

Did you make any changes to the actual text of the original docstring?

ngimel replied:
Rendered docs can be previewed at https://docs-preview.pytorch.org/61002/ (this link is in Dr.CI's comment)

puririshi98 committed:
updated docstring to match and put errors back to warnings to pass the CI

puririshi98 committed:
Merge branch 'autocast_rebase' of https://github.com/puririshi98/pytorch into autocast_rebase

puririshi98 committed:
made tests based on torch.autocast in stead of torch.cuda.amp.autocast

puririshi98 committed:
Update amp.rst

ngimel commented on test/test_cuda.py:
```diff
@@ -2999,7 +2999,7 @@ def test_autocast_rnn(self):
                     h = (h, c)
 
-                with torch.cuda.amp.autocast():
+                with torch.autocast():
```
I think making `torch.autocast` default to cuda is suprising behavior, we should make `device` a required kwarg and not try to infer it.

puririshi98 replied:
okay i will make that change

ngimel commented on torch/autocast_mode.py:
```diff
@@ -0,0 +1,193 @@
+
+    Args:
+        enabled(bool, optional, default=True)":  Whether autocasting should be enabled in the region.
+        fast_dtype(torch_dtype, optional):  Wether to use torch.float16 or torch.bfloat16
```
whether, same on next line

puririshi98 replied:
My mistake! Will correct, thanks for catching it

ngimel commented on torch/cuda/amp/autocast_mode.py:
```diff
@@ -1,146 +1,14 @@
-                return func(*args, **kwargs)
-        return decorate_autocast
 
+autocast = functools.partial(torch.autocast, device_type='cuda', fast_dtype=torch.float16)
```
this enables surprising behavior where `torch.cuda.amp.autocast(device_type='cpu') is valid. Instead you can have `torch.cuda.amp.autocast` inherit from `torch.autocast`, and have its `init` not accept `device_type` argument.

```
class autocast(torch.autocast):
    def __init__(self, fast_dtype=torch.float16):
          super().__init__(device_type="cuda", fast_dtype=fast_dtype)
```
You can also put a check here that bfloat16 is not called on an architecture before ampere, and documentation link pointing to torch.autocast (because we still want torch.cuda.amp.autocast to be documented).

ngimel reviewed (commented):
Also, test errors are real

puririshi98 commented on torch/cuda/amp/autocast_mode.py:
```diff
@@ -1,146 +1,14 @@
-        enabled(bool, optional, default=True):  Whether autocasting should be enabled in the region.
-    """
-    def __init__(self, enabled=True):
-        if enabled and amp_definitely_not_available():
```
I am simply using it because that was what was used previously. If you think it better I can replace it with (not torch.cuda.is_available)

ngimel replied:
Yeah, it used to be needed when xla amp was sharing some code with cuda amp, and it was fragile even at the time. Now xla amp is broken completely for unrelated to your PR reasons, so we can go back to `torch.cuda.is_available`, and when xla is fixed, we'll figure out what to do. cc @mcarilli

puririshi98 committed:
fixing spelling of whether

puririshi98 committed:
fixing inheritance for cuda and cpu amp. Also added a check for cuda devices to only use bfloat16 if ampere architecture or newer

puririshi98 committed:
added new bindings to allow list

puririshi98 committed:
Merge branch 'master' into autocast_rebase

puririshi98 committed:
fixing test usage

puririshi98 committed:
comments for torch.cuda.amp.autocast

puririshi98 committed:
fixing a minor issue

puririshi98 committed:
Update autocast_mode.py

puririshi98 committed:
Update autocast_mode.py

puririshi98 committed:
Update autocast_mode.py

puririshi98 committed:
Update amp.rst

puririshi98 committed:
Update amp.rst

puririshi98 committed:
Update amp.rst

puririshi98 committed:
Update autocast_mode.py

puririshi98 committed:
Update autocast_mode.py

puririshi98 commented:
@ngimel

puririshi98 commented:
@ngimel I have addressed your comments and cleaned up the docs. Let me know if there is anything else needed to get this PR merged.

> 💚 💚 Looks good so far! There are no failures yet. 💚 💚

ngimel commented on torch/autocast_mode.py:
```diff
@@ -0,0 +1,193 @@
+        if self.device == 'cuda':
+            if self.fast_dtype == torch.bfloat16 and torch.cuda.get_device_properties(torch.cuda.current_device()).major < 8:
+                warnings.warn('Current CUDA Device does not support bfloat16. Switching fast_dtype to float16.')
+                self.fast_dtype = torch.float16
```
Don't silently switch dtype, better error out here

ngimel commented on torch/cpu/amp/autocast_mode.py:
```diff
@@ -1,36 +1,5 @@
-            with self:
-                return func(*args, **kwargs)
-        return decorate_autocast
+class autocast(torch.autocast):
```
can you please also add link to documentation here?

puririshi98 replied:
> This looks good, I left minor comments and we can land.

Addressed the comments in the latest commits, let me know if there are any other edits I can make.

ngimel replied:
```suggestion
class autocast(torch.autocast_mode.autocast):
```

ngimel reviewed (commented):
This looks good, I left minor comments and we can land.

puririshi98 committed:
Update autocast_mode.py

puririshi98 committed:
Update autocast_mode.py

ngimel reviewed (approved):
Looks good, let's wait for CI

puririshi98 commented:
@ngimel, most CI passes and the few CI that don't seem to be unrelated to my changes. Are these issues I should address or can this PR be merged.

ngimel commented:
CI failures are unrelated, I'll try landing, thanks!

ngimel commented:
Sorry, it triggers some internal errors, so it'll take some time to fix and land.

puririshi98 commented:
Hi @ngimel, any updates on landing the PR? Is there any way I can help out?

ngimel commented:
Sorry, I was on vacation, will get back to it.

ngimel commented:
@puririshi98 
1) somehow new autocast_mode.py file got windows newlines (^M) instead of linux ones, please fix that
2) inheritance from `torch.autocast` trips pyre typechecker internally, can you please make torch.cuda.amp.autocast and torch.cpu.amp.autocast inherit from `torch.autocast_mode.autocast`?

puririshi98 committed:
fixing inheritance

puririshi98 committed:
Update autocast_mode.py

puririshi98 committed:
replacing newlines

puririshi98 committed:
Update autocast_mode.py

puririshi98 committed:
used notepad++ to switch EOL to linux

puririshi98 commented:
@ngimel I addressed the issues you brought up. Let me know if there are any other changes needed

ngimel commented on torch/cpu/amp/autocast_mode.py:
```diff
@@ -1,6 +1,6 @@
 import torch
 
-class autocast(torch.autocast):
+class autocast(torch.autocast_mode.autocast):
```
can you please do the same thing in torch/cuda/amp/autocast_mode.py?

JamesDeAntonis commented:
Is this done?  I've been monitoring this PR for weeks and just noticed the checks disappeared.

ngimel commented:
It's in the master branch https://github.com/pytorch/pytorch/commits/master

leslie-fang-intel commented:
@puririshi98 Thanks for the PR. I think this PR has changed the default dtype of `torch.cpu.amp.autocast` from `torch.bfloat16` into `torch.float16`.  I am intended to change it back to `torch.bfloat16` since CPU doesn't support `float16` yet.https://github.com/pytorch/pytorch/blob/f8a84a80cdd15aaea12147d7cb35d199aa302dc5/torch/cpu/amp/autocast_mode.py#L8


And this changes should break the UT `test/test_autocast.py` in preCI. But it doesn't. @ngimel Did I make miss some to enable `test/test_autocast.py` in the preCI? such as add `test/test_autocast.py` into `test/run_test.py`.

stas00 commented on torch/autocast_mode.py:
```diff
@@ -0,0 +1,192 @@
+        c_float32 = torch.rand((8, 8), device="cpu")
+        d_float32 = torch.rand((8, 8), device="cpu")
+
+        with autocast(fast_dtype=torch.bfloat16, device_type="cpu"):
```
Is it too late to ask to drop `fast_` in `fast_dtype` - it doesn't feel like an intuitive/functional argument name. 

Is there something wrong with `dtype`? 

Surely it's obvious that this is a target `dtype` for autocast?

Apologies if I missed a discussion on that.

ngimel replied:
No, not too late, let's add it to the request. It used to be `dtype` for cpu APIs, so can be `dtype` here also.

stas00 replied:
Awesome!

Could you please add it, or if not please let me know which request to add it to - thank you!

ngimel replied:
I'm talking to nvidia about it.

Hecim1984 reviewed (commented):
💪🏻🚀
